File: Section-19 Singly Linked Lists/singly_linked_list.py
from typing import Union,Optional

class Node:
    def __init__(self, value=None):
        self.value = value
        self.next = None
    
    def __str__(self):
        return f"{self.value}, {self.next}"


class Singly_Linked_List:
    empty_repr = "Empty"
    def __init__(self, value: Optional[Union[int, float, str]] = None) -> None:
        """Initiating the singly linked list"""
        self.head = Node(value)
        self.tail = self.head

        self.length = 0
        if value!=None:
            self.length = 1
    
    def __len__(self) -> int:
        return self.length
    
    def __iter__(self):
        iterator = next(self)
        return iterator
    
    def __next__(self):
        if self.length==0:
            return []
        temp = self.head
        while temp!=None:
            yield temp.value
            temp = temp.next
    
    def __repr__(self) -> str:
        return_string = ""
        temp = self.head
        while temp!=None:
            return_string += f"{temp.value} -> "
            temp = temp.next
        return_string = return_string[:-4]
        return return_string if return_string!="None" else self.empty_repr
    
    
    def append(self, value: Union[int, float, str]) -> None:
        """Inserts a Node in the tail and reassign the tail to the Node"""
        if value==None:
            raise ValueError("Inserting None type value is not allowed")

        if self.length==0:
            self.__init__(value)
            return

        new_node = Node(value)
        self.tail.next = new_node
        self.tail = new_node
        self.length += 1
    
    def pop(self) -> Union[int, float, str]:
        """Removes the last Node in found in the head and returns the value of the node"""
        if self.length == 0:
            raise ValueError("Can't pop from and empty Singly_Linked_List")
        
        elif self.length == 1:
            value = self.head.value
            self.__init__()
            return value
        
        temp = self.head
        while temp.next.next:
            temp = temp.next
        
        value = temp.next.value
        temp.next = None
        self.length -= 1
        return value

a = Singly_Linked_List()

File: Section-19 Singly Linked Lists/singly_linked_list.js
class Node{
    constructor(value){
        this.value = value
        this.next = undefined
    }
}


class Singly_Linked_List{
    empty_repr = "Empty"
    constructor(value){
        this.head = new Node(value)
        this.tail = this.head

        this.length = 0
        if (this.value !== undefined){
            this.length = 1
        }
    }

    show(){
        let string = ""
        let temp = this.head
        while(temp!==undefined){
            string += `${temp.value} -> `
            temp = temp.next
        }
        
        string = string.slice(0,string.length-4)
        return (string==='undefined') ? this.empty_repr : string
    }

    push(value){
        if (value===undefined){
            throw TypeError("Inserting undefined values are not allowed")
        }
        let new_node = new Node(value)

        if (this.length==0){
            this.head = new_node
            this.tail = this.head
            this.length = 1
            return
        }

        this.tail.next = new_node
        this.tail = new_node
        this.length += 1
    }

    pop(){
        if (this.length===0){
            return undefined
        }
        else if (this.length===1){
            value = this.head.value
            this.head = new Node()
            this.tail = this.head
            this.length = 0
            return value
        }

        let temp = this.head
        while (temp.next.next!==undefined){
            temp = temp.next
        }

        let value = temp.next.value
        this.tail = temp
        temp.next = undefined
        this.length -= 1
        return value
    }
}




//test
a = new Singly_Linked_List()

// for (i=0;i<20;i++){
//     a.push(i)
// }

console.log(a.show())